Fix for "Route not found" issue in custom controllers using pagerfanta

wimvds opened the pull request:
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | 
| Related issues/PRs | 
| License | MIT
| Documentation PR |

#### What's in this PR?

A basic route caching implementation to make getRouteByName work.

#### Why?

While integrating (already existing) custom controllers that used pagerfanta for paging data, we always got a "Route not found" exception. The root cause seemed to be that the getRouteByName function was not defined in the ContentRouteProvider. This PR just caches the route collection for the current request, so the route can be found again when the pagerfanta widget tries to generate it (with paging parameters added).

wimvds committed:
Fix route not found issue on pages with pagerfanta

wachterjohannes reviewed (approved):
LGTM and this specific usecase seems to work also for me. but I am not 100% sure if this is a good idea, because that solves only a quite specific problem.

Maybe that could be a part of solving the problem of generating routes by name. @danrot @chirimoya what do you think? It would be also OK for me to merge this as it is.

wimvds commented:
I'm fully aware that it only fixes this specific use case (ie. anything that uses the route of the current request) and we could also override the provider in our own projects for now. But I think this could help a lot of users that are integrating existing code. No hard feelings if it doesn't get merged though. Still loving Sulu :+1:.

BTW If you were to decide to merge it, would it be ok if I add the same fix to the Sulu\Component\CustomUrl\Routing\CustomUrlRouteProvider as well?

wachterjohannes commented:
@wimvds if we merge it, it would be no problem to add it to all places we need it :) one thing we talked internal today is that we should add a comment to the code why we have added this and which usecase will be fixed by this code. Would you do that?

wimvds commented:
Sure, I'll do it later today or tomorrow.

danrot reviewed (approved):
For me it is also ok to merge. Better support a single use case than no usecase at all.

wimvds committed:
Add fix in CustomUrlRouteProvider & add notice about limitation of getHostByName

wachterjohannes reviewed (approved):
@danrot can be merged in my opinion

wimvds commented:
I noticed that there's still an issue with this : even though it works on the website, the preview in the back-end does not work. I'll try to find out why.

wachterjohannes commented:
@wimvds will you fix this preview bug also?

wimvds commented:
I'm looking into it. Will keep you updated.

wimvds commented:
Nope, I kind of lost track of this issue (we did the easy "fix" and decided to not show the pager on previews). Will have a look at it today, IIRC I was stuck with the routing in the preview kernel.

wimvds commented:
Ran into dependency hell yesterday when trying to upgrade the project we had the issue with (it uses Sylius as well ...). So I decided to create a dummy project (https://github.com/wimvds/pager-sulu-demo) to find a better way to resolve the issue.

wimvds commented:
The root cause of the issue seems to be that the ```sulu_route.routing.router``` is not added in the preview kernel. Trying to figure out why.

wimvds committed:
Hack to make previews work again :p

wimvds commented:
Ok, the root cause seemed to be that the Symfony routing is completely bypassed in the PreviewRenderer. This results in no _route being defined in the request info, and thus causes the "None of the chained routers were able to generate route: Route '' not found, Route 'Symfony\Cmf\Component\Routing\ProviderBasedGenerator' not found" exception. 

So I added a _route in the getByEntity result of the PageRouteDefaultsProvider and added a (dirty) hack to the PreviewRenderer to inject the route (with the actual path of the content).

wimvds committed:
Adapt tests to handle routing hack ...

danrot commented on src/Sulu/Component/CustomUrl/Routing/CustomUrlRouteProvider.php:
```diff
@@ -180,4 +189,24 @@ private function decodePathInfo($pathInfo)
+    /**
+     * @param RouteCollection $collection
+     */
+    public function setRouteCollection(RouteCollection $collection)
```
I don't think it is a good idea to introduce this method, if you only need it for tests... Then I would even prefer to set it via reflections in the test. If we introduce this method people might start using it for whatever reason...

danrot commented on src/Sulu/Bundle/WebsiteBundle/Routing/ContentRouteProvider.php:
```diff
@@ -282,4 +291,24 @@ private function decodePathInfo($pathInfo)
+    /**
+     * @param RouteCollection $collection
+     */
+    public function setRouteCollection(RouteCollection $collection)
```
Same as below, would even prefer reflections in the test if this is the only reason.

danrot reviewed (changes requested):
There is still a CI error... Can you also try to rebase? I guess that the reason for the circle ci error.
@wachterjohannes Can you also have another look?

wimvds committed:
Fix tests

wimvds commented:
Done. Should I also squash the commits or leave them as is?

wachterjohannes committed:
Merge branch 'develop' into cache-route-collection

danrot commented:
@wimvds I am really sorry, seem to have lost track of this PR... Is this still something you want to get merged?

niklasnatter commented:
There are a lot of merge conflicts and the last commit on this was in 2017, therefore I am going to close this. Feel free to open a new PR if this is still a problem - thanks a lot for your time!
